hera_cal/red_groups.py

```diff
@@ -0,0 +1,407 @@
+
+        # Reset the maps
+        obj.clear_cache()
+
+        if not inplace:
+            return obj
+
+    def insert(self, pos: int, red: Sequence[AntPair | Baseline], inplace: bool=True) -> None:
+        """In-place insert a new redundant group to the list of redundant groups.
+        
+        This maintains the list-of-lists duck-typing of the redundant groups.
+        """
+        return self.append(red, inplace=inplace, pos=pos)
+    
+    def __len__(self):
+        return len(self._red_list)
+
+    def __iter__(self):
+        return iter(self._red_list)
+    
+    def __getitem__(self, key: int | AntPair | Baseline) -> list[AntPair | Baseline]:
+        if isinstance(key, int):
+            return self._red_list[key]
+        else:
+            return self.get_red(key)
+        
+    def __setitem__(self, key: int | AntPair | Baseline, value: list[AntPair | Baseline]):
+        if isinstance(key, int):
+            self._red_list[key] = value
+        elif key in self:
+            ukey = self.get_ubl_key(key)
+            self._red_list = [value if red[0]==ukey else red for red in self._red_list]
+        else:
+            # We're setting a new redundant group
+            self.append(value)
+
+        # Reset the maps
+        self.clear_cache()
+
+    def index(self, key: AntPair | Baseline) -> int:
+        """Return the index of a redundant group."""
+        return list(self._red_key_to_bls_map.keys()).index(self.get_ubl_key(key))
+    
+    def filtered(self, **kwargs) -> RedundantGroups:
+        """Return a new RedundantGroups object with baselines filtered out.
+
+        """
+        new_reds = redcal.filter_reds(copy.deepcopy(self), **kwargs)
+        return attrs.evolve(self, red_list=new_reds)
+    
+    def extend(self, reds: Sequence[Sequence[AntPair | Baseline]], inplace: bool = True) -> None:
+        """In-place extend the list of redundant groups with another list of redundant groups.
+        
+        This maintains the list-of-lists duck-typing of the redundant groups.
+        """
+        for red in reds:
+            self.append(red, inplace)
+    
+    def sort(self):
+        """Sort the redundant groups in-place."""
+        self._red_list.sort()
+        self.clear_cache()
+
+    def clear_cache(self):
+        """Clear the cached maps."""
+        for att in (
+            self._bl_to_red_map,
+            self._red_key_to_bls_map,
+            self.data_ants,
+            self.data_bls,
+        ):
+            try:
+                del att
+            except AttributeError:
+                pass
+
+    def get_full_redundancies(
+        self, pols: Sequence[str] | None = None, 
+        include_autos: bool = True, 
+        bl_error_tol: float = 1.0,
+        pol_mode: str='1pol',
+    ) -> RedundantGroups:
+        """Create a RedundantGroups object from an antpos dictionary."""
+        if pols is None:
+            if len(next(iter(self._bl_to_red_map.values()))) == 2:
```

Comment:
Is it possible for this dict to be a mix of 2-tuples and 3-tuples, making the outcome of this command uncertain?

Reply:
No.